src/main/java/com/iota/iri/hash/Kerl.java

```diff
@@ -6,142 +6,122 @@
+            trits[pos + HASH_LENGTH - 1] = 0;
+            bytesFromBigInt(bigIntFromTrits(trits, pos, HASH_LENGTH), state);
+            keccak.update(state);
+        }
     }
 
     @Override
-    public void squeeze(final int[] trits, int offset, int length) {
-
+    public void squeeze(final int[] trits, final int offset, final int length) {
         if (length % 243 != 0) {
-            throw new RuntimeException("Illegal length: " + length);
+            throw new IllegalArgumentException("Illegal length: " + length);
         }
-
         try {
-          do {
-              this.keccak.digest(byte_state, 0, BYTE_HASH_LENGTH);
-              //convert to trits
-              tritsFromBigInt(bigIntFromBytes(byte_state, 0, BYTE_HASH_LENGTH), trit_state, 0, Sponge.HASH_LENGTH);
-
-              //copy with offset
-              trit_state[HASH_LENGTH - 1] = 0;
-              System.arraycopy(trit_state, 0, trits, offset, HASH_LENGTH);
+            for (int pos = offset; pos < offset + length; pos += HASH_LENGTH) {
 
-              //calculate hash again
-              for (int i = byte_state.length; i-- > 0; ) {
+                byte[] state = new byte[BYTE_HASH_LENGTH];
+                keccak.digest(state, 0, BYTE_HASH_LENGTH);
 
-                  byte_state[i] = (byte) (byte_state[i] ^ 0xFF);
-              }
-              keccak.update(byte_state);
-              offset += HASH_LENGTH;
+                //convert into trits
+                BigInteger value = new BigInteger(state);
+                tritsFromBigInt(value, trits, pos, Sponge.HASH_LENGTH);
+                trits[pos + HASH_LENGTH - 1] = 0;
 
-          } while ((length -= HASH_LENGTH) > 0);
+                //calculate hash again
+                for (int i = state.length; i-- > 0; ) {
+                    state[i] = (byte) (state[i] ^ 0xFF);
+                }
+                keccak.update(state);
+            }
         } catch (DigestException e) {
-            e.printStackTrace();
+            e.printStackTrace(System.err);
             throw new RuntimeException(e);
         }
     }
 
     public static BigInteger bigIntFromTrits(final int[] trits, final int offset, final int size) {
-
+        for (int i = offset; i < offset + size; i++) {
+            if (trits[i] < -1 || trits[i] > 1) {
+                throw new IllegalArgumentException("not a trit: " + trits[i]);
+            }
+        }
         BigInteger value = BigInteger.ZERO;
-
-        for (int i = size; i-- > 0; ) {
-
-            value = value.multiply(BigInteger.valueOf(Converter.RADIX)).add(BigInteger.valueOf(trits[offset + i]));
+        for (int n = offset + size - 1; n >= offset; ) {
+            int count = 0;
+            long num = 0L;
+            while (n >= offset && count < MAX_POWERS_LONG) {
+                num = 3 * num + trits[n--];
+                count++;
+            }
+            value = value.multiply(RADIX_POWERS[count]).add(BigInteger.valueOf(num));
         }
```

Comment:
```
BigInteger value = BigInteger.ZERO;
for (int n = offset + size - 1; n >= offset; ) {
    long num = 0L;
    int power = 0;
    while (n >= offset && power < MAX_POWERS_LONG) {
        num = 3 * num + trits[n--];
        power++;
    }
    value = value.multiply(RADIX_POWERS[power]).add(BigInteger.valueOf(num));
}
```
It all comes down to this routine. Do the math the native way ... until the power is too big. Then merge it into the BigInteger ... and do it all again. The radix powers from 0 to 40 are pre-calculated.